[873] Migrate to Poetry

hendzeld opened the pull request:
Resolves #873.

grzanka commented:
@hendzeld add poetry to dependabot as well

hendzeld commented:
> @hendzeld add poetry to dependabot as well

If I read this correctly, it would mean the current configuration is ok.

(https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#package-ecosystem)
![obraz](https://github.com/yaptide/yaptide/assets/128628266/c7ad46aa-6ac9-400d-8c75-fcafd6305ac5)


However, I copied the dependencies as they were specified in requirements-*.txt files.
So they probably should be relaxed to "*" or `major releases` (^Y.x.x) and specific versions should be specified in poetry.lock.

I also need to configure/remove cache from Docker images.
I will also take a look at caching in CI pipelines.

https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#about-caching-workflow-dependencies

grzanka commented:
> > @hendzeld add poetry to dependabot as well
> 
> If I read this correctly, it would mean the current configuration is ok.
> 
> (https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#package-ecosystem) ![obraz](https://private-user-images.githubusercontent.com/128628266/309493516-c7ad46aa-6ac9-400d-8c75-fcafd6305ac5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk0MDgxMzYsIm5iZiI6MTcwOTQwNzgzNiwicGF0aCI6Ii8xMjg2MjgyNjYvMzA5NDkzNTE2LWM3YWQ0NmFhLTZhYzktNDAwZC04Yzc1LWZjYWZkNjMwNWFjNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMwMlQxOTMwMzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zYzlkN2NkY2NlNTcxOWY3YTRmYzc0MTFkZWMyNmE0MjNlMTYwMGNhYzM2ZmFiMGRjM2YxZmYyZGFmOTcxYWVhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.24KjClj98rFWCC_R0hCafn3RNWph10u2Ho0lp5CQOuo)

That good !

> However, I copied the dependencies as they were specified in requirements-*.txt files. So they probably should be relaxed to `*` or `major releases` (^Y.x.x). Then specific versions should be specified and updated in poetry.lock.

I'd freeze some of the packages as they are now, i.e. we haven't migrated the code to flask==3.0. We can fix the flask at 2.x. I am curious if dependabot would then propose a bump to 3.x or will it only propose bumps at stuff defined in the lock file.

> I also need to configure/remove the cache from Docker images. I will also take a look at caching in CI pipelines. (https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#about-caching-workflow-dependencies)

I think that can be left for subsequent PR.

grzanka commented on .gitpod.yml:
```diff
@@ -7,9 +7,11 @@ tasks:
     init: |
-      pip install -r requirements.txt
+      export POETRY_VIRTUALENVS_CREATE=false
+      export FLASK_SQLALCHEMY_DATABASE_URI=sqlite://
```
this seems like in-memory database, I think that file-based may be more universal for gitpod. Easier to inspect.

grzanka commented on Dockerfile-flask:
```diff
@@ -7,9 +8,11 @@ EXPOSE 6000
-RUN pip install --no-cache-dir -r requirements.txt
+# Copy poetry configuration files
+COPY poetry.lock poetry.toml pyproject.toml ./
+# Insatll poatry and project dependencies, we also disables virtualenv creation
```
```suggestion
# Install poetry and project dependencies, we also disabled virtualenv creation
```

grzanka commented on Dockerfile-worker:
```diff
@@ -22,9 +23,11 @@ ENV TOPAS_G4_DATA_DIR="/simulators/geant4_files_path"
-RUN pip install --no-cache-dir -r requirements.txt
+# Copy poetry configuration files
+COPY poetry.lock poetry.toml pyproject.toml ./
+# Insatll poatry and project dependencies, we also disables virtualenv creation
```
fix grammar

hendzeld replied:
Will be changed to one suggested above

```suggestion
# Install poetry and project dependencies, we also disabled virtualenv creation
```

grzanka commented:
@hendzeld let me know once I could test it in yap-dev machine

hendzeld commented:
> > However, I copied the dependencies as they were specified in requirements-*.txt files. So they probably should be relaxed to `*` or `major releases` (^Y.x.x). Then specific versions should be specified and updated in poetry.lock.
> 
> I'd freeze some of the packages as they are now, i.e. we haven't migrated the code to flask==3.0. We can fix the flask at 2.x. I am curious if dependabot would then propose a bump to 3.x or will it only propose bumps at stuff defined in the lock file.

@grzanka 
I created a new repository with a copied configuration. (https://github.com/hendzeld/poetry-dependabot)

Dependabot can update both lock and pyproject files.
I checked documentation and for test I changed its configuration. So it ignores major updates.

Repository: (https://github.com/hendzeld/poetry-dependabot)

grzanka commented on Dockerfile-flask:
```diff
@@ -1,4 +1,5 @@
 FROM python:3.10-slim
+ARG POETRY_VERSION=1.2.0
```
Why 1.2.0 ?
https://pypi.org/project/poetry/#history

hendzeld replied:
Right, I will update it to the latest version.
Probably copy/paste mistake.

grzanka reviewed (changes requested):
https://github.com/yaptide/yaptide/pull/874#issuecomment-1976104488

hendzeld commented:
@grzanka Should work now.

Works for me :)
